python-sdk/pachyderm_sdk/api/pps/extension.py

```diff
@@ -148,3 +155,129 @@ def create_secret(
+        >>> for datum in datums:
+        >>>     print(datum)
+        >>>
+        >>> more_datums = cdc.next_batch(number=10)
+        >>> for datum in more_datums:
+        >>>     print(datum)
+        """
+
+        def __init__(
+            self,
+            create_datum: Callable[[Iterable["CreateDatumRequest"]], Iterator["DatumInfo"]],
+        ):
+            self.queue = SimpleQueue()
+            self.response_it = create_datum(self)
+            self.done = False
+
+        def first_batch(self, input: "Input", number: int) -> List["DatumInfo"]:
+            """Returns the first batch of datums for the provided input spec.
+            Must be called once before `next_batch`.
+
+            Parameters
+            ----------
+            input : pps.Input
+                The input spec.
+            number : int
+                The number of datums to return. If 0, the default batch size set
+                server-side is returned.
+
+            Returns
+            -------
+            List[pps.DatumInfo]
+                List length may be less than `number` if the server has returned
+                all datums.
+            """
+            if self.done:
+                return []
+
+            self.queue.put(
+                CreateDatumRequest(
+                    start=StartCreateDatumRequest(input=input, number=number)
+                )
+            )
+
+            return self._collect_datums(number)
+
+        def next_batch(self, number: int) -> List["DatumInfo"]:
+            """Returns subsequent batches of datums. `first_batch` must have
+            been called first.
+
+            Parameters
+            ----------
+            number : int
+                The number of datums to return. If 0, the default batch size set
+                server-side is returned.
+
+            Returns
+            -------
+            List[pps.DatumInfo]
+                List length may be less than `number` if the server has returned
+                all datums.
+            """
+            if self.done:
+                return []
+
+            self.queue.put(
+                CreateDatumRequest(continue_=ContinueCreateDatumRequest(number=number))
+            )
+
+            return self._collect_datums(number)
+
+        def _collect_datums(self, number: int) -> List["DatumInfo"]:
+            dis = []
+
+            for di in self.response_it:
+                dis.append(di)
+                if len(dis) == number:
+                    break
+
+            # Server has returned all datums
+            if len(dis) < number:
+                self.done = True
+
+            return dis
+
+        def __next__(self):
+            if self.done:
+                # Server returned all datums so terminate request iterator
+                raise StopIteration
+
+            return self.queue.get()
+
+        def __iter__(self):
+            return self
+
+    def create_datum(self) -> "CreateDatumClient":
```

Comment:
create_datum probably means something different to users as opposed to a pachyderm developer. Do we want to think about renaming this to something more self-explanatory? Even `generate_datums` seems a bit better to me

Reply:
Having a different function name for the helper function is good with me.